iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergAcidUtil.java

```diff
@@ -252,13 +314,59 @@ public boolean hasNext() {
     public T next() {
       T next = currentIterator.next();
       GenericRecord rec = (GenericRecord) next;
+      IcebergAcidUtil.copyFields(rec, FILE_READ_META_COLS.size(), current.size(), current);
       PositionDeleteInfo.setIntoConf(conf,
           IcebergAcidUtil.parseSpecId(rec),
           IcebergAcidUtil.computePartitionHash(rec),
           IcebergAcidUtil.parseFilePath(rec),
-          IcebergAcidUtil.parseFilePosition(rec));
-      IcebergAcidUtil.copyFields(rec, FILE_READ_META_COLS.size(), current.size(), current);
+          IcebergAcidUtil.parseFilePosition(rec),
+          StringUtils.EMPTY);
       return (T) current;
     }
   }
+
+  public static class MergeTaskVirtualColumnAwareIterator<T> implements CloseableIterator<T> {
+
+    private final CloseableIterator<T> currentIterator;
+    private GenericRecord current;
+    private final Schema expectedSchema;
+    private final Configuration conf;
+    private final PartitionSpec partitionSpec;
+    private final StructLike partition;
+
+    public MergeTaskVirtualColumnAwareIterator(CloseableIterator<T> currentIterator,
+                                           Schema expectedSchema, Configuration conf, ContentFile contentFile,
+                                           Table table) {
+      this.currentIterator = currentIterator;
+      this.expectedSchema = expectedSchema;
+      this.conf = conf;
+      this.partition = contentFile.partition();
+      current = GenericRecord.create(
+              new Schema(expectedSchema.columns().subList(0, expectedSchema.columns().size())));
+      this.partitionSpec = table.specs().get(contentFile.specId());
+    }
+
+    @Override
+    public void close() throws IOException {
+      currentIterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return currentIterator.hasNext();
+    }
+
+    @Override
+    public T next() {
+      T next = currentIterator.next();
+      GenericRecord rec = (GenericRecord) next;
+      IcebergAcidUtil.setSpecId(current, partitionSpec.specId());
```

Comment:
should we create GenericRecordBuilder
````
new GenericRecordBuilder(schema)
   .withPartitionHash
   .withFilePath
   .withFilePosition
   .build()
````

Reply:
This can be done in some other PR. I dont want this to be included in a single PR.

Reply:
ok, could we at least refactor to
````
IcebergAcidUtil.setPositionDeleteInfo(current, partitionSpec, computeHash(partition), rec)
````

Reply:
We are setting the information to the record itself. Not sure how position delete info came into the picture.

Reply:
aren't we constructing here delete record?

Reply:
No, we are passing the delete information to be deleted directly.

Reply:
i don't catch this. isn't the constructed record used only by HiveIcebergDeleteWriter?

Reply:
`setPositionDeleteInfo` - sets in a current record delete info

Reply:
Refactored to include GenericRecordBuilder. Done.